### PR TITLE
[SPARK-41214][SQL] Fix AQE cache does not update plan and metrics

### PR DESCRIPTION
**Pull request by ulysses-you**
### What changes were proposed in this pull request?

This pr fixs two issues when cache enable AQE
1. Make AQE propagate the SQL metrics which is in InMemoryTableScan
   We do not propagate SQL metrics if the current execution id is not mapping to current query execution. The AdaptiveSparkPlanExec in InMemoryTableScan does not have its own query execution id. So we missed that SQL metrics.
 A simaple case is:
```
          ...
           |
  AdaptiveSparkPlanExec (queryexecution 0, no execution id)
           |
  InMemoryTableScanExec
           |
          ...
           |
  AdaptiveSparkPlanExec (queryexecution 1, execution id 0)
```

2. Update final plan if contains InMemoryTableScan
   We only update final plan if contains subquery to avoid unnecessary `SparkListenerSQLAdaptiveExecutionUpdate` event, however, the AdaptiveSparkPlanExec missed updating if the final stage incude InMemoryTableScan.
 A simaple case is:
```
          ...
           |
  AdaptiveSparkPlanExec
           |
  InMemoryTableScanExec
           |
      ProjectExec
           |
  AdaptiveSparkPlanExec
```

### Why are the changes needed?

Correct the plan and metrics if cache enable AQE. And make Spark UI work. 

### Does this PR introduce _any_ user-facing change?

yes, after this pr, the Spark UI show the correct plan and metrics with AQE cache.

### How was this patch tested?

add test and manually test

### CODE REVIEW

**Comment by ulysses-you** [comment]
cc @erenavsarogullari @cloud-fan @maryannxue

**Comment by erenavsarogullari** [comment]
Thanks @ulysses-you for this fix.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala [hunk 1]
```diff
@@ -2693,6 +2694,21 @@ class AdaptiveQueryExecSuite
...
+      val arr = Seq(
+        (1, "Employee_1", "Department_1"),
+        (2, "Employee_2", "Department_2"))
+      val df = arr.toDF("id", "name", "department").filter($"id" < 3).groupBy($"name").count()
+      df.cache().createOrReplaceTempView("v1")
+      val arr2 = Seq((1, "Employee_1", "Department_1"))
+      val df2 = arr2.toDF("id", "name", "department").filter($"id" > 0).groupBy($"name").count()
+      df2.cache().createOrReplaceTempView("v2")
+
+      runAdaptiveAndVerifyResult("SELECT * FROM v1 JOIN v2 on v1.name = v2.name")
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala:2709]: does this test check metrics?

Comment (ulysses-you) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala:2730]: added one more assert with `SparkListenerSQLAdaptiveSQLMetricUpdates`

**Review by erenavsarogullari — APPROVED** [review]

**Comment by erenavsarogullari** [comment]
Hi @ulysses-you and @cloud-fan,
Hope you are fine.
Do we plan to merge this patch? Thanks.

**Comment by ulysses-you** [comment]
close this in favor of https://github.com/apache/spark/pull/39624

**Pull request closed without merging**